flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTask.java

```diff
@@ -98,13 +96,8 @@ public void init() throws Exception {
 	}
 
 	@Override
-	protected void run() throws Exception {
-		// cache processor reference on the stack, to make the code more JIT friendly
-		final StreamInputProcessor<IN> inputProcessor = this.inputProcessor;
-
-		while (running && inputProcessor.processInput()) {
-			// all the work happens in the "processInput" method
-		}
+	protected boolean performDefaultAction() throws Exception {
```

Comment:
I'm worried about the possible performance degradation caused by moving the `while (true)` loop to the outside level and adding one more virtual call. @sunhaibotb was reporting that this is causing performance degradations. 

Before merging this, can we measure this? Maybe this is visible in `benchmarkCount` or we could write a micro benchmark for the `StreamTask` or `OneInputStreamTask`?

Reply:
Maybe there is an impact, maybe not. But I also don't see any more performant alternative if we want to change the threading model. Besides, I executed at least the `ManualSpeedTest` for now and at least there was no regression.

Reply:
`ManualWindowSpeedITCase` is not necessarily a good benchmark and definitely it's obsolete. Please check out for regressions in our existing `flink-benchmarks` results (once they will be fixed), since I'm pretty sure they supersede this one and/or add an additional one for better `OneInputStreamTask` coverage. We will need those benchmarks sooner or later before we complete this story. The sooner we have them, the better trends will be visible once we finally optimise something (or cause some regression).

Reply:
=== TwoInputXStreamTask.java ===
@Override
protected void run() throws Exception {
	// cache processor reference on the stack, to make the code more JIT friendly
	final StreamTwoInputXProcessor<IN1, IN2> inputProcessor = this.inputProcessor;
	...
	inputProcessor.processInput();
	...
}


=== StreamTwoInputXProcessor.java ===
public boolean processInput() throws Exception {
	while (running) {
		...
		if (element.isRecord()) {
			continue;
		}
		...
	}
	return true;
}


=== TwoInputXStreamTask.java ===

@Override
protected void run() throws Exception {
	// cache processor reference on the stack, to make the code more JIT friendly
	final StreamTwoInputXProcessor<IN1, IN2> inputProcessor = this.inputProcessor;
	...
	while (running && inputProcessor.processInput()) {
		// all the work happens in the "processInput" method
	}
	...
}

@Override
protected void cancelTask() {
	running = false;
}

=== StreamTwoInputXProcessor.java ===
public boolean processInput() throws Exception {
	while (true) {
		...
		if (element.isRecord()) {
			return true;
		}
		...
	}
}

Reply:
=== TwoInputXStreamTask.java ===
@Override
protected void run() throws Exception {
	// cache processor reference on the stack, to make the code more JIT friendly
	final StreamTwoInputXProcessor<IN1, IN2> inputProcessor = this.inputProcessor;
	...
	inputProcessor.processInput();
	...
}


=== StreamTwoInputXProcessor.java ===
public boolean processInput() throws Exception {
	while (running) {
		...
		if (element.isRecord()) {
			continue;
		}
		...
	}
	return true;
}


=== TwoInputXStreamTask.java ===

@Override
protected void run() throws Exception {
	// cache processor reference on the stack, to make the code more JIT friendly
	final StreamTwoInputXProcessor<IN1, IN2> inputProcessor = this.inputProcessor;
	...
	while (running && inputProcessor.processInput()) {
		// all the work happens in the "processInput" method
	}
	...
}

@Override
protected void cancelTask() {
	running = false;
}

=== StreamTwoInputXProcessor.java ===
public boolean processInput() throws Exception {
	while (true) {
		...
		if (element.isRecord()) {
			return true;
		}
		...
	}
}

Reply:
I only tested the following two cases,  Case 1 is much faster than Case 2. I can not sure if the change here has a performance regression.

Case 1:
```
=== TwoInputXStreamTask.java ===
@Override
protected void run() throws Exception {
	// cache processor reference on the stack, to make the code more JIT friendly
	final StreamTwoInputXProcessor<IN1, IN2> inputProcessor = this.inputProcessor;

	inputProcessor.processInput();
}


=== StreamTwoInputXProcessor.java ===
public boolean processInput() throws Exception {
	while (running) {
		...
		if (element.isRecord()) {
			continue;
		}
		...
	}
	return true;
}
```

Case 2:
```
=== TwoInputXStreamTask.java ===

@Override
protected void run() throws Exception {
	// cache processor reference on the stack, to make the code more JIT friendly
	final StreamTwoInputXProcessor<IN1, IN2> inputProcessor = this.inputProcessor;

	while (running && inputProcessor.processInput()) {
		// all the work happens in the "processInput" method
	}
}

=== StreamTwoInputXProcessor.java ===
public boolean processInput() throws Exception {
	while (true) {
		...
		if (element.isRecord()) {
			return true;
		}
		...
	}
}
```

Reply:
I ran the following:
```
		env.setParallelism(1);
		source
			.map(new MultiplyByTwo())
			.addSink(new DiscardingSink<>());

master:
Benchmark                          Mode  Cnt      Score     Error   Units
SumLongsBenchmark.benchmarkCount  thrpt   30  16240.866 ± 311.845  ops/ms

master + #8442:
Benchmark                          Mode  Cnt      Score     Error   Units
SumLongsBenchmark.benchmarkCount  thrpt   30  16540.537 ± 295.100  ops/ms
```

```
		env.setParallelism(1);
		source
			.map(new MultiplyByTwo())
			.rebalance()
			.map((x) -> x)
			.addSink(new DiscardingSink<>());

master:
Benchmark                          Mode  Cnt     Score     Error   Units
SumLongsBenchmark.benchmarkCount  thrpt   30  9715.334 ± 170.632  ops/ms

master + #8442:
Benchmark                          Mode  Cnt     Score     Error   Units
SumLongsBenchmark.benchmarkCount  thrpt   30  9932.752 ± 124.494  ops/ms
```

I would conclude the changes run as fast or a notch faster than master on my local machine.

Reply:
In theory, I don't think it's possible to get faster because I don't see the optimization that might get faster in this change, at most it is the same as before. Feeling some kind of interference makes benchmark inaccurate, or this benchmarking can not accurately measure their performance. In my experience, on the contrary, it will make one more virtual call than before (as @pnowojski  said), because JIT do not inline `performDefaultAction()` which  has more than one implementations. Of course, the accurate benchmark results should be the ultimately criterion.

Reply:
@sunhaibotb if you are in doubt, feel free to checkout the PR and double check the measurement :-) Does your theory also include that the code does a couple more things differently, e.g. it is not checking the `isRunning` volatile read in each iteration? Whether or not inlining or other optimizations for low cardinality polymorphism happend is also unclear, and last but not least the method call could be insignificant in the larger context.

Reply:
I am not even claiming it is faster, probably it is just the same ballpark plus some variance, but at least seemingly no regression to worry about.

Reply:
Btw I am also talking about the performance of #8442, not just this PR.

Reply:
> last but not least the method call could be insignificant in the larger context.

I quite agree with that.

If we narrow down the context and just start to measure from the loop (like the benchmark of [PR-8124](https://github.com/apache/flink/pull/8124)), the difference should show up. But considering in the context of a job,  especially for most production jobs，this difference may have only a slight impact and can be ignored. So it's fine to me. @pnowojski, what do you think?